src/theory/datatypes/theory_datatypes_utils.h

```diff
@@ -178,12 +178,36 @@ Node applySygusArgs(const Datatype& dt,
                     Node op,
                     Node n,
                     const std::vector<Node>& args);
-/**
- * Get the builtin sygus operator for constructor term n of sygus datatype
- * type. For example, if n is the term C_+( d1, d2 ) where C_+ is a sygus
- * constructor whose sygus op is the builtin operator +, this method returns +.
+/** Sygus to builtin
+ *
+ * This method converts a constant term of SyGuS datatype type to its builtin
+ * equivalent. For example, given input C_*( C_x(), C_y() ), this method returns
+ * x*y, assuming C_+, C_x, and C_y have sygus operators *, x, and y
+ * respectively.
+ */
+Node sygusToBuiltin(Node c);
+/** Sygus to builtin eval
+ *
+ * This method returns the rewritten form of (DT_SYGUS_EVAL n args). Notice that
+ * n does not necessarily need to be a constant.
+ *
+ * It does so by (1) converting constant subterms of n to builtin terms and
+ * evaluating them on the arguments args, (2) unfolding non-constant
+ * applications of sygus constructors in n with respect to args and (3)
+ * converting all other non-constant subterms of n to applications of
+ * DT_SYGUS_EVAL.
+ *
+ * For example, if
+ *   n = C_+( C_*( C_x(), C_y() ), n' ), and args = { 3, 4 }
+ * where n' is a variable, then this method returns:
+ *   12 + (DT_SYGUS_EVAL n' 3 4)
```

Comment:
Currently we can't have that `n'` has a different type, right?

Reply:
n' can have arbitrary type wrt n.

Reply:
Isn't it the case then that `n'` can be applied to only some of the arguments that `n` was applied to?

Reply:
No, `DT_SYGUS_EVAL` is a parametric operator; the key is that n and n' are terms pertaining to the same function to synthesize by construction, hence their evaluation operators (on paper) take the same set of arguments.

Reply:
Well, then it's not arbitrary. I thought this was the case, since we cannot define functions-to-synthesize in term of others.